README.md

```diff
@@ -4,7 +4,24 @@
 [![Build Status](https://travis-ci.org/JuliaComputing/MKL.jl.svg?branch=master)](https://travis-ci.org/JuliaComputing/MKL.jl)
 [![Build status](https://ci.appveyor.com/api/projects/status/n37h9eagmnx1gly0/branch/master?svg=true)](https://ci.appveyor.com/project/andreasnoack/mkl-jl/branch/master)
 
-*MKL.jl* is a package that makes Julia's linear algebra use Intel MKL BLAS and LAPACK instead of OpenBLAS. The build step of the package will automatically download Intel MKL and rebuild Julia's system image against Intel MKL. Once the install has completed, you'll have
+*MKL.jl* is a package that makes Julia's linear algebra use Intel MKL BLAS and LAPACK instead of OpenBLAS. The build step of the package will automatically download Intel MKL and rebuild Julia's system image against Intel MKL. 
+
+## To Install:
+
+```julia
+julia>]add https://github.com/JuliaComputing/MKL.jl
+```
+After building which takes some time restart Julia.
+
+If you get the `Intel MKL FATAL ERROR: Cannot load mkl_intel_thread.dll.` error, that means the required dlls are not on the path. You need to add the location that it was installed to your path. 
+On windows usually that is `C:\Users\%USERNAME%\.julia\packages\MKL\36eeJ\deps\usr\bin`. You can add it to your windows path by running the following commmand in cmd:
```

Comment:
That hash 36eeJ may be version specific, or who knows.  Consider suggesting 
```julia
using MKL
joinpath(pathof(MKL), "deps", "usr", "bin")
```
To get the path instead?= for copying
to find the path instead

Reply:
This might not work because the error prevents the REPL from running after the restart. Maybe, before restart it can be done (?!)

Reply:
This is bad; we should never be in a situation where we are unable to load the REPL.

If you can reproduce this error, I would like to inspect the system to see why this is.  Ideally, we would push that directory onto the path automatically.

Reply:
You can reproduce the error by running this on Windows:

```julia
julia>]add https://github.com/JuliaComputing/MKL.jl
```

Reply:
I don't have convenient access to a windows machine at the moment;

Can you show me the contents of `deps/deps.jl` after building `MKL.jl`?

Reply:
This is in my working Julia 1.2 + MKL, which I had added that Intel dll file before building. I can remove that from the path and try building it again if it matters
```julia
## This file autogenerated by BinaryProvider.write_deps_file().
## Do not edit.
##
## Include this file within your main top-level source, and call
## `check_deps()` from within your module's `__init__()` method

if isdefined((@static VERSION < v"0.7.0-DEV.484" ? current_module() : @__MODULE__), :Compat)
    import Compat.Libdl
elseif VERSION >= v"0.7.0-DEV.3382"
    import Libdl
end
const libmkl_core = joinpath(dirname(@__FILE__), "usr\\bin\\mkl_core.dll")
const libmkl_rt = joinpath(dirname(@__FILE__), "usr\\bin\\mkl_rt.dll")
const libmkl_intel_thread = joinpath(dirname(@__FILE__), "usr\\bin\\mkl_intel_thread.dll")
function check_deps()
    global libmkl_core
    if !isfile(libmkl_core)
        error("$(libmkl_core) does not exist, Please re-run Pkg.build(\"MKL\"), and restart Julia.")
    end

    if Libdl.dlopen_e(libmkl_core) in (C_NULL, nothing)
        error("$(libmkl_core) cannot be opened, Please re-run Pkg.build(\"MKL\"), and restart Julia.")
    end

    global libmkl_rt
    if !isfile(libmkl_rt)
        error("$(libmkl_rt) does not exist, Please re-run Pkg.build(\"MKL\"), and restart Julia.")
    end

    if Libdl.dlopen_e(libmkl_rt) in (C_NULL, nothing)
        error("$(libmkl_rt) cannot be opened, Please re-run Pkg.build(\"MKL\"), and restart Julia.")
    end

    global libmkl_intel_thread
    if !isfile(libmkl_intel_thread)
        error("$(libmkl_intel_thread) does not exist, Please re-run Pkg.build(\"MKL\"), and restart Julia.")
    end

    if Libdl.dlopen_e(libmkl_intel_thread) in (C_NULL, nothing)
        error("$(libmkl_intel_thread) cannot be opened, Please re-run Pkg.build(\"MKL\"), and restart Julia.")
    end

end

```

Reply:
Hmmm, this looks to me like it should always work; the path to `libmkl_intel_thread` should be specific enough that `PATH` manipulation should not be needed.

Do you need to add to `PATH` on that computer to get it to work?  If not, do you have access to a computer where you do need to do that in order for it to work?

Reply:
Even If I add it after compilation it works too, but after compilation, the REPL doesn't load anymore, and I need to use Windows shell. I am guessing if we can add the files to the path before compilation that will solve the issue.

Reply:
I think that this code runs after something internal to MKL that needs the path of that dll file @staticfloat

Reply:
What do you mean "you need to use windows shell"?  You're saying that if you double-click on the Julia icon, nothing happens (or you get some kind of error?  If so, what is the error?) but if you open up a Command prompt and then run Julia, you can execute Julia?  What happens if you open up a Command prompt, and run Julia without setting PATH?  Do you get an error?

Reply:
Running REPL only:
![mklREPL](https://user-images.githubusercontent.com/16418197/66081907-a15d1880-e52e-11e9-820f-9138b41686f0.gif)

Inside Atom using Juno:
![mklAtom](https://user-images.githubusercontent.com/16418197/66082035-e2edc380-e52e-11e9-8feb-83d293c3720b.gif)

Reply:
I have to use windows shell to add the folder of ` mkl_intel_thread.dll` to the path manually after MKL building is done and I have restarted my REPL (which doesn't load). If I add the folder to the path it loads

Reply:
Excellent; this is great information, thank you.  This helps me to understand the problem; I'm looking in to how we can do this automatically.

Reply:
@aminya can you try out [this branch](https://github.com/JuliaComputing/MKL.jl/tree/sf/improve_installation_experience) and see if it helps at all?  Also, if you can test out running `MKL.enable_openblas_startup()` afterwards to see if it successfully resets to OpenBLAS, that would be great.

Reply:
@staticfloat unfortunately, the directory was not added to the Windows path again. The same thing happens.
Here is the content of the package without bin and download folder:
https://drive.google.com/file/d/1BIFaXJ2qUW8WTlp9P0xgAuMyx2iOYkfk/view?usp=sharing

Reply:
The most important thing is to know what changes it makes to the `base/sysimg.jl`, `base/Base.jl` and `base/build_h.jl` files.  The changes I made should attempt to add a line of code to the `__init__()` function located in either your `base/sysimg.jl` or `base/Base.jl` files.  Can you look and see if it indeed made the modification?

Reply:
I can't find `base/` that you are talking about. Where should I find it?

Reply:
Great!  Do you see how it's trying to insert an `ENV["PATH"] = ...` into there?

Reply:
This is the whole julia folder under Julia1.3/share. The only thing which is updated is `build_h.jl`
https://drive.google.com/file/d/1Q0Gvy64KhQR-ss4VxGTvtJdcYfyjNjwT/view?usp=sharing

Reply:
Okay, I've added a few more debugging `@info`'s.  Update the branch, then manually run:

```
using MKL
include(joinpath(dirname(dirname(Base.pathof(MKL))), "deps", "deps.jl"))
MKL.enable_mkl_startup(libmkl_rt)
```

And show me what the output is.  You can use [gist](https://gist.github.com) if you don't want to google drive everything.

Reply:
Ran on a new Julia 1.3 RC3 installation.
```julia

julia> using MKL
[ Info: Precompiling MKL [33e6dc65-8f57-5167-99aa-e5a354878fb2]

julia> using LinearAlgebra

julia> BLAS.vendor()
:openblas64

(v1.3) pkg> add https://github.com/JuliaComputing/MKL.jl#sf/improve_installation_experience
  Updating registry at `C:\Users\yahyaaba\.julia\registries\General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
  Updating git-repo `https://github.com/JuliaComputing/MKL.jl`
  Updating git-repo `https://github.com/JuliaComputing/MKL.jl`
 Resolving package versions...
  Updating `C:\Users\yahyaaba\.julia\environments\v1.3\Project.toml`
  [33e6dc65] ~ MKL v0.0.0 #sf/improve_installation_experience (https://github.com/JuliaComputing/MKL.jl)
  Updating `C:\Users\yahyaaba\.julia\environments\v1.3\Manifest.toml`
  [33e6dc65] ~ MKL v0.0.0 #sf/improve_installation_experience (https://github.com/JuliaComputing/MKL.jl)
  Building MKL → `C:\Users\yahyaaba\.julia\packages\MKL\ocvyd\deps\build.log`

julia> using MKL

julia> include(joinpath(dirname(dirname(Base.pathof(MKL))), "deps", "deps.jl"))
check_deps (generic function with 1 method)

julia> MKL.enable_mkl_startup(libmkl_rt)
[ Info: Found init function in sysimg.jl
┌ Error: Error occured while running user line editor:
│   e = PCRE compilation error: unrecognized character follows \ at offset 30
└ @ MKL C:\Users\yahyaaba\.julia\packages\MKL\BqnHJ\src\install.jl:27
┌ Info: Building `inference.o`:
└ `'C:\Julia-1.3.0-rc3\bin\julia.exe' -C native --output-ji 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.ji' --output-o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.o' compiler/compiler.jl`
essentials.jl
ctypes.jl
generator.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
pair.jl
traits.jl
range.jl
expr.jl
error.jl
bool.jl
number.jl
int.jl
operators.jl
pointer.jl
refvalue.jl
indices.jl
array.jl
abstractarray.jl
bitarray.jl
bitset.jl
abstractdict.jl
abstractset.jl
iterators.jl
namedtuple.jl
docs/core.jl
ordering.jl
sort.jl
compiler/utilities.jl
compiler/validation.jl
compiler/inferenceresult.jl
compiler/params.jl
compiler/inferencestate.jl
compiler/typeutils.jl
compiler/typelimits.jl
compiler/typelattice.jl
compiler/tfuncs.jl
compiler/abstractinterpretation.jl
compiler/typeinfer.jl
compiler/optimize.jl
compiler/ssair/driver.jl
compiler/ssair/ir.jl
compiler/ssair/domtree.jl
compiler/ssair/slot2ssa.jl
compiler/ssair/queries.jl
compiler/ssair/passes.jl
compiler/ssair/inlining.jl
compiler/ssair/verify.jl
compiler/ssair/legacy.jl
compiler/bootstrap.jl
┌ Info: Building `sys.o`:
└ `'C:\Julia-1.3.0-rc3\bin\julia.exe' -C native --output-ji 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.ji' --output-o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.o' -J 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.ji' --startup-file=no sysimg.jl`
coreio.jl
exports.jl
essentials.jl
ctypes.jl
gcutils.jl
generator.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
expr.jl
pair.jl
traits.jl
range.jl
error.jl
bool.jl
number.jl
int.jl
operators.jl
pointer.jl
refvalue.jl
refpointer.jl
checked.jl
indices.jl
array.jl
abstractarray.jl
subarray.jl
views.jl
baseext.jl
ntuple.jl
abstractdict.jl
iterators.jl
namedtuple.jl
hashing.jl
rounding.jl
float.jl
twiceprecision.jl
complex.jl
rational.jl
multinverses.jl
abstractarraymath.jl
arraymath.jl
simdloop.jl
reduce.jl
reshapedarray.jl
reinterpretarray.jl
bitarray.jl
bitset.jl
multimedia.jl
some.jl
dict.jl
abstractset.jl
set.jl
char.jl
strings/basic.jl
strings/string.jl
strings/substring.jl
build_h.jl
version_git.jl
osutils.jl
c.jl
io.jl
iobuffer.jl
intfuncs.jl
strings/strings.jl
strings/search.jl
strings/unicode.jl
strings/util.jl
strings/io.jl
parse.jl
shell.jl
regex.jl
pcre.jl
show.jl
arrayshow.jl
methodshow.jl
cartesian.jl
multidimensional.jl
permuteddimsarray.jl
broadcast.jl
missing.jl
version.jl
sysinfo.jl
libc.jl
env.jl
linked_list.jl
condition.jl
threads.jl
lock.jl
task.jl
weakkeydict.jl
logging.jl
libuv.jl
uv_constants.jl
asyncevent.jl
iostream.jl
stream.jl
filesystem.jl
cmd.jl
process.jl
grisu/grisu.jl
secretbuffer.jl
floatfuncs.jl
math.jl
reducedim.jl
accumulate.jl
ordering.jl
sort.jl
fastmath.jl
Enums.jl
gmp.jl
mpfr.jl
combinatorics.jl
hashing2.jl
irrationals.jl
mathconstants.jl
printf.jl
meta.jl
channels.jl
deepcopy.jl
download.jl
summarysize.jl
errorshow.jl
stacktraces.jl
initdefs.jl
threadcall.jl
uuid.jl
loading.jl
util.jl
asyncmap.jl
experimental.jl
deprecated.jl
docs\basedocs.jl
client.jl
docs\Docs.jl
Base  ─────────── 25.495423 seconds
Base64  ─────────  3.732911 seconds
CRC32c  ─────────  0.010316 seconds
SHA  ────────────  0.180690 seconds
FileWatching  ───  0.091285 seconds
Unicode  ────────  0.009353 seconds
Mmap  ───────────  0.075999 seconds
Serialization  ──  0.871307 seconds
Libdl  ──────────  0.031111 seconds
Markdown  ───────  1.055684 seconds
LibGit2  ────────  2.662726 seconds
Logging  ────────  0.520022 seconds
Sockets  ────────  1.688433 seconds
Printf  ─────────  0.014142 seconds
Profile  ────────  0.157649 seconds
Dates  ──────────  2.226413 seconds
DelimitedFiles  ─  0.099404 seconds
Random  ─────────  0.515376 seconds
UUIDs  ──────────  0.020330 seconds
Future  ─────────  0.010100 seconds
LinearAlgebra  ── 12.227116 seconds
SparseArrays  ───  6.318208 seconds
SuiteSparse  ────  1.857242 seconds
Distributed  ────  6.729605 seconds
SharedArrays  ───  0.306738 seconds
Pkg  ──────────── 13.893923 seconds
Test  ───────────  0.994980 seconds
REPL  ───────────  1.193482 seconds
Statistics  ─────  0.179796 seconds
Stdlibs total  ── 83.466055 seconds
Sysimage built. Summary:
Total ─────── 108.964538 seconds
Base: ───────  25.495423 seconds 23.3979%
Stdlibs: ────  83.466055 seconds 76.5993%
Build shared library "C:\\Users\\yahyaaba\\.julia\\packages\\PackageCompiler\\CJQcs\\sysimg\\backup\\native\\sys.dll":
  `'C:\Users\yahyaaba\.julia\packages\WinRPM\Y9QdZ\deps\usr\x86_64-w64-mingw32\sys-root\mingw\bin\gcc.exe' --sysroot 'C:\Users\yahyaaba\.julia\packages\WinRPM\Y9QdZ\deps\usr\x86_64-w64-mingw32\sys-root' -shared '-DJULIAC_PROGRAM_LIBNAME="C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll"' -o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll' -Wl,--whole-archive 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.o' -Wl,--no-whole-archive -std=gnu99 '-IC:\Julia-1.3.0-rc3\include\julia' -DJULIA_ENABLE_THREADING=1 '-LC:\Julia-1.3.0-rc3\bin' -Wl,--stack,8388608 -ljulia -lopenlibm -m64 -Wl,--export-all-symbols`
ERROR: IOError: unlink: permission denied (EACCES)
Stacktrace:
 [1] #rm at .\none:0 [inlined]
 [2] copy_system_image(::String, ::String, ::Bool) at C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\src\PackageCompiler.jl:29
 [3] copy_system_image at C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\src\PackageCompiler.jl:19 [inlined]
 [4] force_native_image! at C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\src\api.jl:124 [inlined]
 [5] force_native_image! at C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\src\api.jl:123 [inlined]
 [6] enable_mkl_startup(::String) at C:\Users\yahyaaba\.julia\packages\MKL\BqnHJ\src\install.jl:129
 [7] top-level scope at REPL[8]:1
```
After installation, I reset the REPL, and it doesn't load again
share folder under Julia-1.3.0-RC3:
https://drive.google.com/file/d/1PHhcN2LTiVYzDotZK5YSI4yhYvlWDlLY/view?usp=sharing
MKL package folder without bin and download:
https://drive.google.com/file/d/1mb5xM_30xid6-1nTJ2NHdqVeE4fCp_MZ/view?usp=sharing

Reply:
Great, the `PCRE compilation error: unrecognized character follows \ at offset 30` is important, let me look at that....

Reply:
Interesting; what do the following say for you if you paste them into your REPL:

```
lines = ["a", "b", "function __init__()", "c", "end"]
match.(r"^function\s+__init__\(\)\s*$", lines)
match.(r"^function\s+__init__\(\)\s*$", lines)
```

Those lines give me something reasonable on my windows machine; they don't error out.  How about for you?

Reply:
When I paste them manually it works
```julia
julia> lines = ["a", "b", "function __init__()", "c", "end"]
5-element Array{String,1}:
 "a"
 "b"
 "function __init__()"
 "c"
 "end"

julia> match.(r"^function\s+__init__\(\)\s*$", lines)
5-element Array{Union{Nothing, RegexMatch},1}:
 nothing
 nothing
 RegexMatch("function __init__()")
 nothing
 nothing

julia> match.(r"^function\s+__init__\(\)\s*$", lines)
5-element Array{Union{Nothing, RegexMatch},1}:
 nothing
 nothing
 RegexMatch("function __init__()")
 nothing
 nothing
```

Reply:
Can you edit your `MKL.jl/src/install.jl` file, add a `rethrow(e)` right after the `@error()` on line 27 and run `MKL.enable_mkl_startup(libmkl_rt)` again?

Reply:
```julia
julia> using LinearAlgebra

julia> BLAS.vendor()
:openblas64

julia> using MKL
[ Info: Precompiling MKL [33e6dc65-8f57-5167-99aa-e5a354878fb2]

julia> Base.pathof(MKL)
"C:\\Users\\yahyaaba\\.julia\\packages\\MKL\\ocvyd\\src\\MKL.jl"

julia> include(joinpath(dirname(dirname(Base.pathof(MKL))), "deps", "deps.jl"))
check_deps (generic function with 1 method)

julia> MKL.enable_mkl_startup(libmkl_rt)
[ Info: Replacing libblas_name in C:\Julia-1.3.0-rc3\bin\..\share\julia\base\build_h.jl
[ Info: Checking if we need to update PATH...
[ Info: Checking sysimg.jl
[ Info: Could not find init function in sysimg.jl
[ Info: Checking Base.jl
┌ Error: Error occured while running user line editor:
│   e = PCRE compilation error: unrecognized character follows \ at offset 30
└ @ MKL C:\Users\yahyaaba\.julia\packages\MKL\ocvyd\src\install.jl:27
ERROR: PCRE compilation error: unrecognized character follows \ at offset 30
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] compile(::String, ::UInt32) at .\pcre.jl:123
 [3] compile(::Regex) at .\regex.jl:72
 [4] Regex(::String, ::UInt32, ::UInt32) at .\regex.jl:37
 [5] Regex at .\regex.jl:60 [inlined]
 [6] (::MKL.var"#13#14"{String,String})(::Array{String,1}) at C:\Users\yahyaaba\.julia\packages\MKL\ocvyd\src\install.jl:97
 [7] lineedit(::MKL.var"#13#14"{String,String}, ::String) at C:\Users\yahyaaba\.julia\packages\MKL\ocvyd\src\install.jl:25
 [8] force_proper_PATH(::String, ::String) at C:\Users\yahyaaba\.julia\packages\MKL\ocvyd\src\install.jl:87
 [9] enable_mkl_startup(::String) at C:\Users\yahyaaba\.julia\packages\MKL\ocvyd\src\install.jl:127
 [10] top-level scope at REPL[4]:1
```

Reply:
I did some debugging, this is the workspace right after the error:
![image](https://user-images.githubusercontent.com/16418197/66616893-321f9e00-eb98-11e9-80f9-85e29d5ef869.png)

Reply:
Ah, perfect, thanks.  This helps; it appears to be the backslash escaping.  I've pushed a new version that is a little simpler; please try one more time.

Reply:
This time code works! Yay 🚀 🎉 🎈 
```julia


(v1.3) pkg> add https://github.com/JuliaComputing/MKL.jl#sf/improve_installation_experience
  Updating registry at `C:\Users\yahyaaba\.julia\registries\General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
  Updating git-repo `https://github.com/JuliaComputing/MKL.jl`
  Updating git-repo `https://github.com/JuliaComputing/MKL.jl`
 Resolving package versions...
  Updating `C:\Users\yahyaaba\.julia\environments\v1.3\Project.toml`
  [33e6dc65] ~ MKL v0.0.0 #sf/improve_installation_experience (https://github.com/JuliaComputing/MKL.jl)
  Updating `C:\Users\yahyaaba\.julia\environments\v1.3\Manifest.toml`
  [33e6dc65] ~ MKL v0.0.0 #sf/improve_installation_experience (https://github.com/JuliaComputing/MKL.jl)
  Building MKL → `C:\Users\yahyaaba\.julia\packages\MKL\ohxi8\deps\build.log`

julia> using MKL
[ Info: Precompiling MKL [33e6dc65-8f57-5167-99aa-e5a354878fb2]
┌ Warning: The call to compilecache failed to create a usable precompiled cache file for MKL [33e6dc65-8f57-5167-99aa-e5a354878fb2]
│   exception = Required dependency LibGit2 [76f85450-5226-5b5a-8eaa-529ad045b433] failed to load from a cache file.
└ @ Base loading.jl:1036

julia> include(joinpath(dirname(dirname(Base.pathof(MKL))), "deps", "deps.jl"))
check_deps (generic function with 1 method)

julia> MKL.enable_mkl_startup(libmkl_rt)
[ Info: Replacing libblas_name in C:\Julia-1.3.0-rc3\bin\..\share\julia\base\build_h.jl
[ Info: Checking if we need to update PATH...
[ Info: Checking sysimg.jl
[ Info: Could not find init function in sysimg.jl
[ Info: Checking Base.jl
[ Info: Found ENV already
┌ Info: Building `inference.o`:
└ `'C:\Julia-1.3.0-rc3\bin\julia.exe' -C native --output-ji 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.ji' --output-o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.o' compiler/compiler.jl`
essentials.jl
ctypes.jl
generator.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
pair.jl
traits.jl
range.jl
expr.jl
error.jl
bool.jl
number.jl
int.jl
operators.jl
pointer.jl
refvalue.jl
indices.jl
array.jl
abstractarray.jl
bitarray.jl
bitset.jl
abstractdict.jl
abstractset.jl
iterators.jl
namedtuple.jl
docs/core.jl
ordering.jl
sort.jl
compiler/utilities.jl
compiler/validation.jl
compiler/inferenceresult.jl
compiler/params.jl
compiler/inferencestate.jl
compiler/typeutils.jl
compiler/typelimits.jl
compiler/typelattice.jl
compiler/tfuncs.jl
compiler/abstractinterpretation.jl
compiler/typeinfer.jl
compiler/optimize.jl
compiler/ssair/driver.jl
compiler/ssair/ir.jl
compiler/ssair/domtree.jl
compiler/ssair/slot2ssa.jl
compiler/ssair/queries.jl
compiler/ssair/passes.jl
compiler/ssair/inlining.jl
compiler/ssair/verify.jl
compiler/ssair/legacy.jl
compiler/bootstrap.jl
┌ Info: Building `sys.o`:
└ `'C:\Julia-1.3.0-rc3\bin\julia.exe' -C native --output-ji 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.ji' --output-o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.o' -J 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.ji' --startup-file=no sysimg.jl`
coreio.jl
exports.jl
essentials.jl
ctypes.jl
gcutils.jl
generator.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
expr.jl
pair.jl
traits.jl
range.jl
error.jl
bool.jl
number.jl
int.jl
operators.jl
pointer.jl
refvalue.jl
refpointer.jl
checked.jl
indices.jl
array.jl
abstractarray.jl
subarray.jl
views.jl
baseext.jl
ntuple.jl
abstractdict.jl
iterators.jl
namedtuple.jl
hashing.jl
rounding.jl
float.jl
twiceprecision.jl
complex.jl
rational.jl
multinverses.jl
abstractarraymath.jl
arraymath.jl
simdloop.jl
reduce.jl
reshapedarray.jl
reinterpretarray.jl
bitarray.jl
bitset.jl
multimedia.jl
some.jl
dict.jl
abstractset.jl
set.jl
char.jl
strings/basic.jl
strings/string.jl
strings/substring.jl
build_h.jl
version_git.jl
osutils.jl
c.jl
io.jl
iobuffer.jl
intfuncs.jl
strings/strings.jl
strings/search.jl
strings/unicode.jl
strings/util.jl
strings/io.jl
parse.jl
shell.jl
regex.jl
pcre.jl
show.jl
arrayshow.jl
methodshow.jl
cartesian.jl
multidimensional.jl
permuteddimsarray.jl
broadcast.jl
missing.jl
version.jl
sysinfo.jl
libc.jl
env.jl
linked_list.jl
condition.jl
threads.jl
lock.jl
task.jl
weakkeydict.jl
logging.jl
libuv.jl
uv_constants.jl
asyncevent.jl
iostream.jl
stream.jl
filesystem.jl
cmd.jl
process.jl
grisu/grisu.jl
secretbuffer.jl
floatfuncs.jl
math.jl
reducedim.jl
accumulate.jl
ordering.jl
sort.jl
fastmath.jl
Enums.jl
gmp.jl
mpfr.jl
combinatorics.jl
hashing2.jl
irrationals.jl
mathconstants.jl
printf.jl
meta.jl
channels.jl
deepcopy.jl
download.jl
summarysize.jl
errorshow.jl
stacktraces.jl
initdefs.jl
threadcall.jl
uuid.jl
loading.jl
util.jl
asyncmap.jl
experimental.jl
deprecated.jl
docs\basedocs.jl
client.jl
docs\Docs.jl
Base  ─────────── 25.161805 seconds
Base64  ─────────  3.681222 seconds
CRC32c  ─────────  0.010250 seconds
SHA  ────────────  0.182722 seconds
FileWatching  ───  0.091057 seconds
Unicode  ────────  0.009989 seconds
Mmap  ───────────  0.074055 seconds
Serialization  ──  0.834911 seconds
Libdl  ──────────  0.032856 seconds
Markdown  ───────  1.060167 seconds
LibGit2  ────────  2.712762 seconds
Logging  ────────  0.506565 seconds
Sockets  ────────  1.667135 seconds
Printf  ─────────  0.011522 seconds
Profile  ────────  0.153575 seconds
Dates  ──────────  2.083787 seconds
DelimitedFiles  ─  0.102044 seconds
Random  ─────────  0.508742 seconds
UUIDs  ──────────  0.021498 seconds
Future  ─────────  0.012847 seconds
LinearAlgebra  ── 11.354594 seconds
SparseArrays  ───  6.152054 seconds
SuiteSparse  ────  1.822965 seconds
Distributed  ────  6.912383 seconds
SharedArrays  ───  0.220522 seconds
Pkg  ──────────── 12.485028 seconds
Test  ───────────  0.991497 seconds
REPL  ───────────  1.168580 seconds
Statistics  ─────  0.165613 seconds
Stdlibs total  ── 55.074695 seconds
Sysimage built. Summary:
Total ───────  80.240112 seconds
Base: ───────  25.161805 seconds 31.3581%
Stdlibs: ────  55.074695 seconds 68.6374%
Build shared library "C:\\Users\\yahyaaba\\.julia\\packages\\PackageCompiler\\CJQcs\\sysimg\\backup\\native\\sys.dll":
  `'C:\Users\yahyaaba\.julia\packages\WinRPM\Y9QdZ\deps\usr\x86_64-w64-mingw32\sys-root\mingw\bin\gcc.exe' --sysroot 'C:\Users\yahyaaba\.julia\packages\WinRPM\Y9QdZ\deps\usr\x86_64-w64-mingw32\sys-root' -shared '-DJULIAC_PROGRAM_LIBNAME="C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll"' -o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll' -Wl,--whole-archive 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.o' -Wl,--no-whole-archive -std=gnu99 '-IC:\Julia-1.3.0-rc3\include\julia' -DJULIA_ENABLE_THREADING=1 '-LC:\Julia-1.3.0-rc3\bin' -Wl,--stack,8388608 -ljulia -lopenlibm -m64 -Wl,--export-all-symbols`
ERROR: IOError: unlink: permission denied (EACCES)
Stacktrace:
 [1] #rm at .\none:0 [inlined]
 [2] copy_system_image(::String, ::String, ::Bool) at C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\src\PackageCompiler.jl:29
 [3] copy_system_image at C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\src\PackageCompiler.jl:19 [inlined]
 [4] force_native_image! at C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\src\api.jl:124 [inlined]
 [5] force_native_image! at C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\src\api.jl:123 [inlined]
 [6] enable_mkl_startup(::String) at C:\Users\yahyaaba\.julia\packages\MKL\ohxi8\src\install.jl:133
 [7] top-level scope at REPL[5]:1
```

After restart:
```julia
julia> using LinearAlgebra
julia> BLAS.vendor()
:mkl                                                                                                             
```

Reply:
Enabling openBlas also works:
```julia
julia> using MKL

julia> MKL.enable_openblas_startup()
[ Info: Replacing libblas_name in C:\Julia-1.3.0-rc3\bin\..\share\julia\base\build_h.jl
┌ Info: Building `inference.o`:
└ `'C:\Julia-1.3.0-rc3\bin\julia.exe' -C native --output-ji 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.ji' --output-o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.o' compiler/compiler.jl`
essentials.jl
ctypes.jl
generator.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
pair.jl
traits.jl
range.jl
expr.jl
error.jl
bool.jl
number.jl
int.jl
operators.jl
pointer.jl
refvalue.jl
indices.jl
array.jl
abstractarray.jl
bitarray.jl
bitset.jl
abstractdict.jl
abstractset.jl
iterators.jl
namedtuple.jl
docs/core.jl
ordering.jl
sort.jl
compiler/utilities.jl
compiler/validation.jl
compiler/inferenceresult.jl
compiler/params.jl
compiler/inferencestate.jl
compiler/typeutils.jl
compiler/typelimits.jl
compiler/typelattice.jl
compiler/tfuncs.jl
compiler/abstractinterpretation.jl
compiler/typeinfer.jl
compiler/optimize.jl
compiler/ssair/driver.jl
compiler/ssair/ir.jl
compiler/ssair/domtree.jl
compiler/ssair/slot2ssa.jl
compiler/ssair/queries.jl
compiler/ssair/passes.jl
compiler/ssair/inlining.jl
compiler/ssair/verify.jl
compiler/ssair/legacy.jl
compiler/bootstrap.jl
┌ Info: Building `sys.o`:
└ `'C:\Julia-1.3.0-rc3\bin\julia.exe' -C native --output-ji 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.ji' --output-o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.o' -J 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.ji' --startup-file=no sysimg.jl`
coreio.jl
exports.jl
essentials.jl
ctypes.jl
gcutils.jl
generator.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
expr.jl
pair.jl
traits.jl
range.jl
error.jl
bool.jl
number.jl
int.jl
operators.jl
pointer.jl
refvalue.jl
refpointer.jl
checked.jl
indices.jl
array.jl
abstractarray.jl
subarray.jl
views.jl
baseext.jl
ntuple.jl
abstractdict.jl
iterators.jl
namedtuple.jl
hashing.jl
rounding.jl
float.jl
twiceprecision.jl
complex.jl
rational.jl
multinverses.jl
abstractarraymath.jl
arraymath.jl
simdloop.jl
reduce.jl
reshapedarray.jl
reinterpretarray.jl
bitarray.jl
bitset.jl
multimedia.jl
some.jl
dict.jl
abstractset.jl
set.jl
char.jl
strings/basic.jl
strings/string.jl
strings/substring.jl
build_h.jl
version_git.jl
osutils.jl
c.jl
io.jl
iobuffer.jl
intfuncs.jl
strings/strings.jl
strings/search.jl
strings/unicode.jl
strings/util.jl
strings/io.jl
parse.jl
shell.jl
regex.jl
pcre.jl
show.jl
arrayshow.jl
methodshow.jl
cartesian.jl
multidimensional.jl
permuteddimsarray.jl
broadcast.jl
missing.jl
version.jl
sysinfo.jl
libc.jl
env.jl
linked_list.jl
condition.jl
threads.jl
lock.jl
task.jl
weakkeydict.jl
logging.jl
libuv.jl
uv_constants.jl
asyncevent.jl
iostream.jl
stream.jl
filesystem.jl
cmd.jl
process.jl
grisu/grisu.jl
secretbuffer.jl
floatfuncs.jl
math.jl
reducedim.jl
accumulate.jl
ordering.jl
sort.jl
fastmath.jl
Enums.jl
gmp.jl
mpfr.jl
combinatorics.jl
hashing2.jl
irrationals.jl
mathconstants.jl
printf.jl
meta.jl
channels.jl
deepcopy.jl
download.jl
summarysize.jl
errorshow.jl
stacktraces.jl
initdefs.jl
threadcall.jl
uuid.jl
loading.jl
util.jl
asyncmap.jl
experimental.jl
deprecated.jl
docs\basedocs.jl
client.jl
docs\Docs.jl
Base  ─────────── 25.654957 seconds
Base64  ─────────  3.794240 seconds
CRC32c  ─────────  0.010188 seconds
SHA  ────────────  0.181228 seconds
FileWatching  ───  0.091911 seconds
Unicode  ────────  0.010193 seconds
Mmap  ───────────  0.074751 seconds
Serialization  ──  0.862091 seconds
Libdl  ──────────  0.030937 seconds
Markdown  ───────  1.054587 seconds
LibGit2  ────────  2.649440 seconds
Logging  ────────  0.506780 seconds
Sockets  ────────  1.877789 seconds
Printf  ─────────  0.012369 seconds
Profile  ────────  0.153111 seconds
Dates  ──────────  2.098834 seconds
DelimitedFiles  ─  0.104160 seconds
Random  ─────────  0.525311 seconds
UUIDs  ──────────  0.019996 seconds
Future  ─────────  0.011412 seconds
LinearAlgebra  ── 12.891522 seconds
SparseArrays  ───  6.296322 seconds
SuiteSparse  ────  1.802302 seconds
Distributed  ────  6.485953 seconds
SharedArrays  ───  0.222248 seconds
Pkg  ──────────── 13.223301 seconds
Test  ───────────  0.961266 seconds
REPL  ───────────  1.154462 seconds
Statistics  ─────  0.162380 seconds
Stdlibs total  ── 57.315055 seconds
Sysimage built. Summary:
Total ───────  82.973635 seconds
Base: ───────  25.654957 seconds 30.9194%
Stdlibs: ────  57.315055 seconds 69.0762%
Build shared library "C:\\Users\\yahyaaba\\.julia\\packages\\PackageCompiler\\CJQcs\\sysimg\\backup\\native\\sys.dll":
  `'C:\Users\yahyaaba\.julia\packages\WinRPM\Y9QdZ\deps\usr\x86_64-w64-mingw32\sys-root\mingw\bin\gcc.exe' --sysroot 'C:\Users\yahyaaba\.julia\packages\WinRPM\Y9QdZ\deps\usr\x86_64-w64-mingw32\sys-root' -shared '-DJULIAC_PROGRAM_LIBNAME="C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll"' -o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll' -Wl,--whole-archive 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.o' -Wl,--no-whole-archive -std=gnu99 '-IC:\Julia-1.3.0-rc3\include\julia' -DJULIA_ENABLE_THREADING=1 '-LC:\Julia-1.3.0-rc3\bin' -Wl,--stack,8388608 -ljulia -lopenlibm -m64 -Wl,--export-all-symbols`
[ Info: Copying system image: C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll to C:\Julia-1.3.0-rc3\lib\julia\sys.dll


```

Reply:
> 
> 
> Enabling openBlas also works:
> 
> ```julia
> julia> using MKL
> 
> julia> MKL.enable_openblas_startup()
> [ Info: Replacing libblas_name in C:\Julia-1.3.0-rc3\bin\..\share\julia\base\build_h.jl
> ┌ Info: Building `inference.o`:
> └ `'C:\Julia-1.3.0-rc3\bin\julia.exe' -C native --output-ji 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.ji' --output-o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.o' compiler/compiler.jl`
> essentials.jl
> ctypes.jl
> generator.jl
> reflection.jl
> options.jl
> promotion.jl
> tuple.jl
> pair.jl
> traits.jl
> range.jl
> expr.jl
> error.jl
> bool.jl
> number.jl
> int.jl
> operators.jl
> pointer.jl
> refvalue.jl
> indices.jl
> array.jl
> abstractarray.jl
> bitarray.jl
> bitset.jl
> abstractdict.jl
> abstractset.jl
> iterators.jl
> namedtuple.jl
> docs/core.jl
> ordering.jl
> sort.jl
> compiler/utilities.jl
> compiler/validation.jl
> compiler/inferenceresult.jl
> compiler/params.jl
> compiler/inferencestate.jl
> compiler/typeutils.jl
> compiler/typelimits.jl
> compiler/typelattice.jl
> compiler/tfuncs.jl
> compiler/abstractinterpretation.jl
> compiler/typeinfer.jl
> compiler/optimize.jl
> compiler/ssair/driver.jl
> compiler/ssair/ir.jl
> compiler/ssair/domtree.jl
> compiler/ssair/slot2ssa.jl
> compiler/ssair/queries.jl
> compiler/ssair/passes.jl
> compiler/ssair/inlining.jl
> compiler/ssair/verify.jl
> compiler/ssair/legacy.jl
> compiler/bootstrap.jl
> ┌ Info: Building `sys.o`:
> └ `'C:\Julia-1.3.0-rc3\bin\julia.exe' -C native --output-ji 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.ji' --output-o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.o' -J 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\basecompiler.ji' --startup-file=no sysimg.jl`
> coreio.jl
> exports.jl
> essentials.jl
> ctypes.jl
> gcutils.jl
> generator.jl
> reflection.jl
> options.jl
> promotion.jl
> tuple.jl
> expr.jl
> pair.jl
> traits.jl
> range.jl
> error.jl
> bool.jl
> number.jl
> int.jl
> operators.jl
> pointer.jl
> refvalue.jl
> refpointer.jl
> checked.jl
> indices.jl
> array.jl
> abstractarray.jl
> subarray.jl
> views.jl
> baseext.jl
> ntuple.jl
> abstractdict.jl
> iterators.jl
> namedtuple.jl
> hashing.jl
> rounding.jl
> float.jl
> twiceprecision.jl
> complex.jl
> rational.jl
> multinverses.jl
> abstractarraymath.jl
> arraymath.jl
> simdloop.jl
> reduce.jl
> reshapedarray.jl
> reinterpretarray.jl
> bitarray.jl
> bitset.jl
> multimedia.jl
> some.jl
> dict.jl
> abstractset.jl
> set.jl
> char.jl
> strings/basic.jl
> strings/string.jl
> strings/substring.jl
> build_h.jl
> version_git.jl
> osutils.jl
> c.jl
> io.jl
> iobuffer.jl
> intfuncs.jl
> strings/strings.jl
> strings/search.jl
> strings/unicode.jl
> strings/util.jl
> strings/io.jl
> parse.jl
> shell.jl
> regex.jl
> pcre.jl
> show.jl
> arrayshow.jl
> methodshow.jl
> cartesian.jl
> multidimensional.jl
> permuteddimsarray.jl
> broadcast.jl
> missing.jl
> version.jl
> sysinfo.jl
> libc.jl
> env.jl
> linked_list.jl
> condition.jl
> threads.jl
> lock.jl
> task.jl
> weakkeydict.jl
> logging.jl
> libuv.jl
> uv_constants.jl
> asyncevent.jl
> iostream.jl
> stream.jl
> filesystem.jl
> cmd.jl
> process.jl
> grisu/grisu.jl
> secretbuffer.jl
> floatfuncs.jl
> math.jl
> reducedim.jl
> accumulate.jl
> ordering.jl
> sort.jl
> fastmath.jl
> Enums.jl
> gmp.jl
> mpfr.jl
> combinatorics.jl
> hashing2.jl
> irrationals.jl
> mathconstants.jl
> printf.jl
> meta.jl
> channels.jl
> deepcopy.jl
> download.jl
> summarysize.jl
> errorshow.jl
> stacktraces.jl
> initdefs.jl
> threadcall.jl
> uuid.jl
> loading.jl
> util.jl
> asyncmap.jl
> experimental.jl
> deprecated.jl
> docs\basedocs.jl
> client.jl
> docs\Docs.jl
> Base  ─────────── 25.654957 seconds
> Base64  ─────────  3.794240 seconds
> CRC32c  ─────────  0.010188 seconds
> SHA  ────────────  0.181228 seconds
> FileWatching  ───  0.091911 seconds
> Unicode  ────────  0.010193 seconds
> Mmap  ───────────  0.074751 seconds
> Serialization  ──  0.862091 seconds
> Libdl  ──────────  0.030937 seconds
> Markdown  ───────  1.054587 seconds
> LibGit2  ────────  2.649440 seconds
> Logging  ────────  0.506780 seconds
> Sockets  ────────  1.877789 seconds
> Printf  ─────────  0.012369 seconds
> Profile  ────────  0.153111 seconds
> Dates  ──────────  2.098834 seconds
> DelimitedFiles  ─  0.104160 seconds
> Random  ─────────  0.525311 seconds
> UUIDs  ──────────  0.019996 seconds
> Future  ─────────  0.011412 seconds
> LinearAlgebra  ── 12.891522 seconds
> SparseArrays  ───  6.296322 seconds
> SuiteSparse  ────  1.802302 seconds
> Distributed  ────  6.485953 seconds
> SharedArrays  ───  0.222248 seconds
> Pkg  ──────────── 13.223301 seconds
> Test  ───────────  0.961266 seconds
> REPL  ───────────  1.154462 seconds
> Statistics  ─────  0.162380 seconds
> Stdlibs total  ── 57.315055 seconds
> Sysimage built. Summary:
> Total ───────  82.973635 seconds
> Base: ───────  25.654957 seconds 30.9194%
> Stdlibs: ────  57.315055 seconds 69.0762%
> Build shared library "C:\\Users\\yahyaaba\\.julia\\packages\\PackageCompiler\\CJQcs\\sysimg\\backup\\native\\sys.dll":
>   `'C:\Users\yahyaaba\.julia\packages\WinRPM\Y9QdZ\deps\usr\x86_64-w64-mingw32\sys-root\mingw\bin\gcc.exe' --sysroot 'C:\Users\yahyaaba\.julia\packages\WinRPM\Y9QdZ\deps\usr\x86_64-w64-mingw32\sys-root' -shared '-DJULIAC_PROGRAM_LIBNAME="C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll"' -o 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll' -Wl,--whole-archive 'C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.o' -Wl,--no-whole-archive -std=gnu99 '-IC:\Julia-1.3.0-rc3\include\julia' -DJULIA_ENABLE_THREADING=1 '-LC:\Julia-1.3.0-rc3\bin' -Wl,--stack,8388608 -ljulia -lopenlibm -m64 -Wl,--export-all-symbols`
> [ Info: Copying system image: C:\Users\yahyaaba\.julia\packages\PackageCompiler\CJQcs\sysimg\backup\native\sys.dll to C:\Julia-1.3.0-rc3\lib\julia\sys.dll
> ```
> 
> After restart:
> 
> ```julia
> julia> using LinearAlgebra
> 
> julia> BLAS.vendor()
> :openblas64
> ```

Then how to disable OpenBLAS and use MKL again?

Reply:
Do you really need this? I recommend having two Julia installation instead of going back and forth.